sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableServiceImpl.java

```diff
@@ -210,6 +234,246 @@ public Row getCurrentRow() throws NoSuchElementException {
+    private long bufferSizeLimit = DEFAULT_BYTE_LIMIT;
+    private int miniBatchWaterMark;
+    private final String tableName;
+
+    @VisibleForTesting
+    BigtableMiniBatchReaderImpl(BigtableSession session, BigtableSource source) {
+      this.session = session;
+      if (source.getMaxBufferElementCount() != null && source.getMaxBufferElementCount() != 0) {
+        this.miniBatchLimit = source.getMaxBufferElementCount();
+        this.bufferSizeLimit = (long) miniBatchLimit * 100 * 1024 * 1024;
+      }
+      this.miniBatchWaterMark = miniBatchLimit / 10;
+      tableName =
+          session.getOptions().getInstanceName().toTableNameStr(source.getTableId().get());
+      this.source = source;
+    }
+
+    @Override
+    public boolean start() throws IOException {
+      buffer = new ArrayDeque<>();
+      lastFillComplete = false;
+      RowRange[] rowRanges = new RowRange[source.getRanges().size()];
+      for (int i = 0; i < source.getRanges().size(); i++) {
+        rowRanges[i] =
+            RowRange.newBuilder()
+                .setStartKeyClosed(
+                    ByteString.copyFrom(source.getRanges().get(i).getStartKey().getValue()))
+                .setEndKeyOpen(
+                    ByteString.copyFrom(source.getRanges().get(i).getEndKey().getValue()))
+                .build();
+      }
+      // Presort the ranges so that future segmentation can exit early when splitting the row set
+      Arrays.sort(rowRanges, RANGE_START_COMPARATOR);
+      rowSet =
+          RowSet.newBuilder()
+              .addAllRowRanges(Arrays.stream(rowRanges).collect(Collectors.toList()))
+              .build();
+
+      HashMap<String, String> baseLabels = new HashMap<>();
+      baseLabels.put(MonitoringInfoConstants.Labels.PTRANSFORM, "");
+      baseLabels.put(MonitoringInfoConstants.Labels.SERVICE, "BigTable");
+      baseLabels.put(MonitoringInfoConstants.Labels.METHOD, "google.bigtable.v2.ReadRows");
+      baseLabels.put(
+          MonitoringInfoConstants.Labels.RESOURCE,
+          GcpResourceIdentifiers.bigtableResource(
+              session.getOptions().getProjectId(),
+              session.getOptions().getInstanceId(),
+              source.getTableId().get()));
+      baseLabels.put(
+          MonitoringInfoConstants.Labels.BIGTABLE_PROJECT_ID, session.getOptions().getProjectId());
+      baseLabels.put(
+          MonitoringInfoConstants.Labels.INSTANCE_ID, session.getOptions().getInstanceId());
+      baseLabels.put(
+          MonitoringInfoConstants.Labels.TABLE_ID,
+          GcpResourceIdentifiers.bigtableTableID(
+              session.getOptions().getProjectId(),
+              session.getOptions().getInstanceId(),
+              source.getTableId().get()));
+      serviceCallMetric =
+          new ServiceCallMetric(MonitoringInfoConstants.Urns.API_REQUEST_COUNT, baseLabels);
+
+      //future = session.getDataClient().readRowsAsync(buildReadRowsRequest());
+      future = createFuture();
```

Comment:
please remove stale comments and please rename the method to be something more descriptive. ie `loadNextMiniBatchAsync()`

Reply:
Also please be consistent in your naming: segment vs minibatch

Reply:
Done.